<!doctype html>
<html class="no-js" lang="en" data-content_root="./">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Usage Notes" href="notes.html" /><link rel="prev" title="cellarr" href="index.html" />

    <!-- Generated with Sphinx 7.4.7 and Furo 2024.08.06 -->
        <title>Cell Arrays - cellarr 0.2.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">cellarr 0.2.3 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">cellarr 0.2.3 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Cell Arrays</a></li>
<li class="toctree-l1"><a class="reference internal" href="#install-the-package">Install the package</a></li>
<li class="toctree-l1"><a class="reference internal" href="#build-the-cellarrdataset">Build the <code class="docutils literal notranslate"><span class="pre">CellArrDataset</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="#query-a-cellarrdataset">Query a <code class="docutils literal notranslate"><span class="pre">CellArrDataset</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="#a-single-cell-dataloader">A single cell dataloader</a></li>
<li class="toctree-l1"><a class="reference internal" href="notes.html">Usage Notes</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="api/modules.html">Module Reference</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="api/cellarr.html">cellarr package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="_sources/tutorial.md.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section class="tex2jax_ignore mathjax_ignore" id="cell-arrays">
<h1>Cell Arrays<a class="headerlink" href="#cell-arrays" title="Link to this heading">¶</a></h1>
<p>Cell Arrays is a Python package that provides a TileDB-backed store for large collections of
genomic experimental data, such as millions of cells across multiple single-cell experiment objects.</p>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="install-the-package">
<h1>Install the package<a class="headerlink" href="#install-the-package" title="Link to this heading">¶</a></h1>
<p>To get started, install the package from <a class="reference external" href="https://pypi.org/project/cellarr/">PyPI</a></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>cellarr
</pre></div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="build-the-cellarrdataset">
<h1>Build the <code class="docutils literal notranslate"><span class="pre">CellArrDataset</span></code><a class="headerlink" href="#build-the-cellarrdataset" title="Link to this heading">¶</a></h1>
<p>The <code class="docutils literal notranslate"><span class="pre">CellArrDataset</span></code> method is designed to store single-cell RNA-seq
datasets but can be generalized to store any 2-dimensional experimental data.</p>
<p>This method creates four TileDB files in the directory specified by <code class="docutils literal notranslate"><span class="pre">output_path</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">gene_annotation</span></code>: A TileDB file containing feature/gene annotations.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sample_metadata</span></code>: A TileDB file containing sample metadata.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">cell_metadata</span></code>: A TileDB file containing cell metadata including mapping to the samples
they are tagged with in <code class="docutils literal notranslate"><span class="pre">sample_metadata</span></code>.</p></li>
<li><p>A matrix TileDB file named by the <code class="docutils literal notranslate"><span class="pre">layer_matrix_name</span></code> parameter. This allows the package
to store multiple different matrices, e.g. normalized, scaled for the same cell, gene, sample
metadata attributes.</p></li>
</ul>
<p>The TileDB matrix file is stored in a <code class="docutils literal notranslate"><span class="pre">cell</span> <span class="pre">X</span> <span class="pre">gene</span></code> orientation. This orientation
is chosen because the fastest-changing dimension as new files are added to the
collection is usually the cells rather than genes.</p>
<p>Process:</p>
<ol class="arabic simple">
<li><p><strong>Scan the Collection</strong>: Scan the entire collection of files to create
a unique set of feature ids (e.g. gene symbols). Store this set as the
<code class="docutils literal notranslate"><span class="pre">gene_annotation</span></code> TileDB file.</p></li>
<li><p><strong>Sample Metadata</strong>: Store sample metadata in <code class="docutils literal notranslate"><span class="pre">sample_metadata</span></code>
TileDB file. Each file is typically considered a sample, and an automatic
mapping is created between files and samples.</p></li>
<li><p><strong>Store Cell Metadata</strong>: Store cell metadata in the <code class="docutils literal notranslate"><span class="pre">cell_metadata</span></code>
TileDB file.</p></li>
<li><p><strong>Remap and Orient Data</strong>: For each dataset in the collection,
remap and orient the feature dimension using the feature set from Step 1.
This step ensures consistency in gene measurement and order, even if
some genes are unmeasured or ordered differently in the original experiments.</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Check out the <a class="reference external" href="https://biocpy.github.io/cellarr/api/cellarr#module-cellarr.build_cellarrdataset">reference</a> documentation for modifying the parameters for any of these steps.</p>
</div>
<p><img alt=" structure" src="_images/cellarr.png" /></p>
<p>First lets mock a few <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="k">def</span> <span class="nf">generate_adata</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">d</span><span class="p">,</span> <span class="n">k</span><span class="p">):</span>
    <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">z</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">k</span><span class="p">),</span> <span class="n">scale</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">k</span><span class="p">)</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
    <span class="n">w</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">k</span><span class="p">))</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">z</span><span class="p">,</span> <span class="n">w</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

    <span class="n">gene_index</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;gene_</span><span class="si">{</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">d</span><span class="p">)]</span>
    <span class="n">var_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;names&quot;</span><span class="p">:</span> <span class="n">gene_index</span><span class="p">},</span> <span class="n">index</span><span class="o">=</span><span class="n">gene_index</span><span class="p">)</span>
    <span class="n">obs_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;cells&quot;</span><span class="p">:</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;cell1_</span><span class="si">{</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">)]})</span>

    <span class="n">adata</span> <span class="o">=</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">layers</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">y</span><span class="p">},</span> <span class="n">var</span><span class="o">=</span><span class="n">var_df</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">obs_df</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">adata</span>

<span class="n">adata1</span> <span class="o">=</span> <span class="n">generate_adata</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">adata2</span> <span class="o">=</span> <span class="n">generate_adata</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">100</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;datasets&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>datasets
AnnData object with n_obs × n_vars = 1000 × 100
    obs: &#39;cells&#39;
    var: &#39;names&#39;
    layers: &#39;counts&#39; AnnData object with n_obs × n_vars = 100 × 1000
    obs: &#39;cells&#39;
    var: &#39;names&#39;
    layers: &#39;counts&#39;
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/cellarr/cellarr/.tox/docs/lib/python3.9/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
/home/runner/work/cellarr/cellarr/.tox/docs/lib/python3.9/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
</pre></div>
</div>
</div>
</div>
<p>To build a <code class="docutils literal notranslate"><span class="pre">CellArrDataset</span></code> from a collection of <code class="docutils literal notranslate"><span class="pre">H5AD</span></code> or <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> objects:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">from</span> <span class="nn">cellarr</span> <span class="kn">import</span> <span class="n">build_cellarrdataset</span><span class="p">,</span> <span class="n">CellArrDataset</span><span class="p">,</span> <span class="n">MatrixOptions</span>

<span class="c1"># Create a temporary directory</span>
<span class="n">tempdir</span> <span class="o">=</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">mkdtemp</span><span class="p">()</span>

<span class="c1"># # Read AnnData objects</span>
<span class="c1"># adata1 = anndata.read_h5ad(&quot;path/to/object1.h5ad&quot;, &quot;r&quot;)</span>
<span class="c1"># # or just provide the path</span>
<span class="c1"># adata2 = &quot;path/to/object2.h5ad&quot;</span>

<span class="c1"># Build CellArrDataset</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">build_cellarrdataset</span><span class="p">(</span>
    <span class="n">output_path</span><span class="o">=</span><span class="n">tempdir</span><span class="p">,</span>
    <span class="n">files</span><span class="o">=</span><span class="p">[</span><span class="n">adata1</span><span class="p">,</span> <span class="n">adata2</span><span class="p">],</span>
    <span class="n">matrix_options</span><span class="o">=</span><span class="n">MatrixOptions</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">),</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/cellarr/cellarr/.tox/docs/lib/python3.9/site-packages/cellarr/build_cellarrdataset.py:217: UserWarning: Scanning all files for feature ids (e.g. gene symbols), this may take long
  warnings.warn(
/home/runner/work/cellarr/cellarr/.tox/docs/lib/python3.9/site-packages/cellarr/build_cellarrdataset.py:279: UserWarning: Scanning all files to compute cell counts, this may take long
  warnings.warn(
/home/runner/work/cellarr/cellarr/.tox/docs/lib/python3.9/site-packages/cellarr/build_cellarrdataset.py:286: UserWarning: Sample metadata is not provided, each dataset in &#39;files&#39; is considered a sample
  warnings.warn(
/home/runner/work/cellarr/cellarr/.tox/docs/lib/python3.9/site-packages/cellarr/build_cellarrdataset.py:303: UserWarning: Scanning all files for feature ids (e.g. gene symbols), this may take long
  warnings.warn(
/home/runner/work/cellarr/cellarr/.tox/docs/lib/python3.9/site-packages/cellarr/build_cellarrdataset.py:332: UserWarning: Scanning all files to compute cell counts, this may take long
  warnings.warn(
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Optimizing /tmp/tmp4w6vp431/gene_annotation
Fragments before consolidation: 1
Fragments after consolidation: 1
Optimizing /tmp/tmp4w6vp431/sample_metadata
Fragments before consolidation: 1
Fragments after consolidation: 1
Optimizing /tmp/tmp4w6vp431/cell_metadata
Fragments before consolidation: 1
Fragments after consolidation: 1
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Optimizing /tmp/tmp4w6vp431/counts
Fragments before consolidation: 2
Fragments after consolidation: 1
class: CellArrDataset
number_of_rows: 1100
number_of_columns: 1000
path: &#39;/tmp/tmp4w6vp431&#39;
</pre></div>
</div>
</div>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>All files are expected to be consistent and any modifications
to make them consistent is outside the scope of this function
and package.</p>
<p>There’s a few assumptions this process makes:</p>
<ul class="simple">
<li><p>If object in <code class="docutils literal notranslate"><span class="pre">files</span></code> is an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code>
or H5AD object, these must contain an assay matrix in the
layers slot of the object named as <code class="docutils literal notranslate"><span class="pre">layer_matrix_name</span></code> parameter.</p></li>
<li><p>Feature information must contain a column defined by the parameter
<code class="docutils literal notranslate"><span class="pre">feature_column</span></code> in the
<code class="docutils literal notranslate"><span class="pre">GeneAnnotationOptions.</span></code> that
contains feature ids or gene symbols across all files.</p></li>
<li><p>If no <code class="docutils literal notranslate"><span class="pre">cell_metadata</span></code> is provided, we scan to count the number of cells
and create a simple range index.</p></li>
<li><p>Each file is considered a sample and a mapping between cells and samples
is automatically created. Hence the sample information provided must match
the number of input files.</p></li>
</ul>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="query-a-cellarrdataset">
<h1>Query a <code class="docutils literal notranslate"><span class="pre">CellArrDataset</span></code><a class="headerlink" href="#query-a-cellarrdataset" title="Link to this heading">¶</a></h1>
<p>Users have the option to reuse the <code class="docutils literal notranslate"><span class="pre">dataset</span></code> object retuned when building the dataset or by creating a <code class="docutils literal notranslate"><span class="pre">CellArrDataset</span></code> object by initializing it to the path where the files were created.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a CellArrDataset object from the existing dataset</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">CellArrDataset</span><span class="p">(</span><span class="n">dataset_path</span><span class="o">=</span><span class="n">tempdir</span><span class="p">)</span>

<span class="c1"># Query data from the dataset</span>
<span class="n">expression_data</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">[</span><span class="mi">10</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;gene_1&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_10&quot;</span><span class="p">,</span> <span class="s2">&quot;gene_500&quot;</span><span class="p">]]</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;matrix slice:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">expression_data</span><span class="o">.</span><span class="n">matrix</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2"> gene_annotation slice:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">expression_data</span><span class="o">.</span><span class="n">gene_annotation</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2"> cell_metadata slice:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">expression_data</span><span class="o">.</span><span class="n">cell_metadata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>shape: (1, 3)
matrix slice:
  (0, 0)	12.436595916748047
  (0, 1)	22.075590133666992


 gene_annotation slice:
    cellarr_gene_index
0               gene_1
1              gene_10
447           gene_500


 cell_metadata slice:
   cellarr_sample cellarr_cell_index_in_sample
10       sample_1                           10
</pre></div>
</div>
</div>
</div>
<p>This returns a <code class="docutils literal notranslate"><span class="pre">CellArrDatasetSlice</span></code> object that contains the matrix and metadata <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code>’s along the cell and gene axes.</p>
<p>Users can easily convert these to analysis-ready representations</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s2">&quot;as anndata:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">expression_data</span><span class="o">.</span><span class="n">to_anndata</span><span class="p">())</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2"> as summarizedexperiment:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">expression_data</span><span class="o">.</span><span class="n">to_summarizedexperiment</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>as anndata:
AnnData object with n_obs × n_vars = 1 × 3
    obs: &#39;cellarr_sample&#39;, &#39;cellarr_cell_index_in_sample&#39;
    var: &#39;cellarr_gene_index&#39;
    layers: &#39;matrix&#39;


 as summarizedexperiment:
class: SummarizedExperiment
dimensions: (3, 1)
assays(1): [&#39;matrix&#39;]
row_data columns(1): [&#39;cellarr_gene_index&#39;]
row_names(3): [&#39;0&#39;, &#39;1&#39;, &#39;447&#39;]
column_data columns(2): [&#39;cellarr_sample&#39;, &#39;cellarr_cell_index_in_sample&#39;]
column_names(1): [&#39;10&#39;]
metadata(0): 
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/cellarr/cellarr/.tox/docs/lib/python3.9/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
/home/runner/work/cellarr/cellarr/.tox/docs/lib/python3.9/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="a-single-cell-dataloader">
<h1>A single cell dataloader<a class="headerlink" href="#a-single-cell-dataloader" title="Link to this heading">¶</a></h1>
<p>A basic single cell dataloader can be instantiated by using the <code class="docutils literal notranslate"><span class="pre">DataModule</span></code> class.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">cellarr.dataloader</span> <span class="kn">import</span> <span class="n">DataModule</span>

<span class="n">datamodule</span> <span class="o">=</span> <span class="n">DataModule</span><span class="p">(</span>
    <span class="n">dataset_path</span><span class="o">=</span><span class="s2">&quot;/path/to/cellar/dir&quot;</span><span class="p">,</span>
    <span class="n">cell_metadata_uri</span><span class="o">=</span><span class="s2">&quot;cell_metadata&quot;</span><span class="p">,</span>
    <span class="n">gene_annotation_uri</span><span class="o">=</span><span class="s2">&quot;gene_annotation&quot;</span><span class="p">,</span>
    <span class="n">matrix_uri</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span>
    <span class="n">label_column_name</span><span class="o">=</span><span class="s2">&quot;label&quot;</span><span class="p">,</span>
    <span class="n">study_column_name</span><span class="o">=</span><span class="s2">&quot;study&quot;</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">lognorm</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Users can optionally set a list of studies to be used as validation. If not provided, all studies are used for training.
Additionally users may also provide the gene space to train their models.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">val_studies</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;study1&quot;</span><span class="p">,</span> <span class="s2">&quot;study100&quot;</span><span class="p">]</span>

<span class="n">gene_list</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s2">&quot;GPNMB&quot;</span><span class="p">,</span> <span class="s2">&quot;TREM2&quot;</span><span class="p">,</span> <span class="s2">&quot;LPL&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-DQA1&quot;</span><span class="p">,</span> <span class="s2">&quot;CD109&quot;</span><span class="p">,</span>
    <span class="s2">&quot;IL6ST&quot;</span><span class="p">,</span> <span class="s2">&quot;SDC2&quot;</span><span class="p">,</span> <span class="s2">&quot;MSR1&quot;</span><span class="p">,</span> <span class="s2">&quot;ALCAM&quot;</span><span class="p">,</span> <span class="s2">&quot;SLC1A3&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CD9&quot;</span><span class="p">,</span> <span class="s2">&quot;CD59&quot;</span><span class="p">,</span> <span class="s2">&quot;MRC1&quot;</span><span class="p">,</span> <span class="s2">&quot;SLC11A1&quot;</span><span class="p">,</span> <span class="s2">&quot;CPM&quot;</span><span class="p">,</span>
    <span class="s2">&quot;GPR183&quot;</span><span class="p">,</span> <span class="s2">&quot;ITGAX&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-DMB&quot;</span><span class="p">,</span> <span class="s2">&quot;NRP2&quot;</span><span class="p">,</span> <span class="s2">&quot;SV2C&quot;</span><span class="p">,</span>
    <span class="s2">&quot;PTPRJ&quot;</span><span class="p">,</span> <span class="s2">&quot;EMP1&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-DQB1&quot;</span><span class="p">,</span> <span class="s2">&quot;MERTK&quot;</span><span class="p">,</span> <span class="s2">&quot;CD52&quot;</span><span class="p">,</span>
    <span class="s2">&quot;CXCL16&quot;</span><span class="p">,</span> <span class="s2">&quot;ABCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;HLA-DPB1&quot;</span><span class="p">,</span> <span class="s2">&quot;OLR1&quot;</span><span class="p">,</span> <span class="s2">&quot;CD83&quot;</span>
<span class="p">]</span>

<span class="n">datamodule</span> <span class="o">=</span> <span class="n">DataModule</span><span class="p">(</span>
    <span class="n">dataset_path</span><span class="o">=</span><span class="s2">&quot;/path/to/cellar/dir&quot;</span><span class="p">,</span>
    <span class="n">cell_metadata_uri</span><span class="o">=</span><span class="s2">&quot;cell_metadata&quot;</span><span class="p">,</span>
    <span class="n">gene_annotation_uri</span><span class="o">=</span><span class="s2">&quot;gene_annotation&quot;</span><span class="p">,</span>
    <span class="n">matrix_uri</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span>
    <span class="n">val_studies</span><span class="o">=</span><span class="n">val_studies</span><span class="p">,</span>
    <span class="n">label_column_name</span><span class="o">=</span><span class="s2">&quot;label&quot;</span><span class="p">,</span>
    <span class="n">study_column_name</span><span class="o">=</span><span class="s2">&quot;study&quot;</span><span class="p">,</span>
    <span class="n">gene_order</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>
    <span class="n">batch_size</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">lognorm</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Users can access training cells by index.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">datamodule</span><span class="o">.</span><span class="n">train_dataset</span><span class="p">[</span><span class="mi">100</span><span class="p">]</span>
</pre></div>
</div>
<p>Batches can be created and examined.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dataloader</span> <span class="o">=</span> <span class="n">datamodule</span><span class="o">.</span><span class="n">train_dataloader</span><span class="p">()</span>

<span class="n">batch</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">dataloader</span><span class="p">))</span>
<span class="n">expression</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">studies</span> <span class="o">=</span> <span class="n">batch</span>
</pre></div>
</div>
<p>The dataloader can then be used in training models. The package provides a simple autoencoder to serve as a template for the user’s own models.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pytorch_lightning</span> <span class="k">as</span> <span class="nn">pl</span>
<span class="kn">from</span> <span class="nn">cellarr.autoencoder</span> <span class="kn">import</span> <span class="n">AutoEncoder</span>

<span class="n">autoencoder</span> <span class="o">=</span> <span class="n">AutoEncoder</span><span class="p">(</span>
    <span class="n">n_genes</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">datamodule</span><span class="o">.</span><span class="n">gene_indices</span><span class="p">),</span>
    <span class="n">latent_dim</span><span class="o">=</span><span class="mi">128</span><span class="p">,</span>
    <span class="n">hidden_dim</span><span class="o">=</span><span class="p">[</span><span class="mi">1024</span><span class="p">,</span> <span class="mi">1024</span><span class="p">,</span> <span class="mi">1024</span><span class="p">],</span>
    <span class="n">dropout</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
    <span class="n">input_dropout</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span>
    <span class="n">residual</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">model_path</span> <span class="o">=</span> <span class="s2">&quot;/path/to/model/mymodel/&quot;</span>
<span class="n">params</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;max_epochs&quot;</span><span class="p">:</span> <span class="mi">500</span><span class="p">,</span>
    <span class="s2">&quot;logger&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
    <span class="s2">&quot;log_every_n_steps&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
    <span class="s2">&quot;limit_train_batches&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span> <span class="c1"># to specify number of batches per epoch</span>
<span class="p">}</span>
<span class="n">trainer</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">Trainer</span><span class="p">(</span><span class="o">**</span><span class="n">params</span><span class="p">)</span>
<span class="n">trainer</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">autoencoder</span><span class="p">,</span> <span class="n">datamodule</span><span class="o">=</span><span class="n">datamodule</span><span class="p">)</span>
<span class="n">autoencoder</span><span class="o">.</span><span class="n">save_all</span><span class="p">(</span><span class="n">model_path</span><span class="o">=</span><span class="n">model_path</span><span class="p">)</span>
</pre></div>
</div>
<hr class="docutils" />
<p>Check out the <a class="reference external" href="https://biocpy.github.io/cellarr/api/modules.html">documentation</a> for more details.</p>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="notes.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Usage Notes</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Home</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Jayaram Kancherla
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="_static/documentation_options.js?v=a47416a6"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>