<!doctype html>
<html class="no-js" lang="en" data-content_root="../../">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="../../genindex.html" /><link rel="search" title="Search" href="../../search.html" />

    <!-- Generated with Sphinx 7.3.7 and Furo 2024.05.06 -->
        <title>cellarr.build_cellarrdataset - cellarr 0.0.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo.css?v=387cc868" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/furo-extensions.css?v=36a5483c" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" /
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="../../index.html"><div class="brand">cellarr 0.0.3 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="../../index.html">
  
  
  <span class="sidebar-brand-text">cellarr 0.0.3 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="../../search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul>
<li class="toctree-l1"><a class="reference internal" href="../../tutorial.html">Overview</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../api/modules.html">Module Reference</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../api/cellarr.html">cellarr package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../license.html">License</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <h1>Source code for cellarr.build_cellarrdataset</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Build the `CellArrDatset`.</span>

<span class="sd">The `CellArrDataset` method is designed to store single-cell RNA-seq</span>
<span class="sd">datasets but can be generalized to store any 2-dimensional experimental data.</span>

<span class="sd">This method creates four TileDB files in the directory specified by `output_path`:</span>

<span class="sd">- `gene_annotation`: A TileDB file containing feature/gene annotations.</span>
<span class="sd">- `sample_metadata`: A TileDB file containing sample metadata.</span>
<span class="sd">- `cell_metadata`: A TileDB file containing cell metadata including mapping to the samples</span>
<span class="sd">they are tagged with in ``sample_metadata``.</span>
<span class="sd">- A matrix TileDB file named by the `layer_matrix_name` parameter. This allows the package</span>
<span class="sd">to store multiple different matrices, e.g. normalized, scaled for the same cell, gene, sample</span>
<span class="sd">metadata attributes.</span>

<span class="sd">The TileDB matrix file is stored in a ``cell X gene`` orientation. This orientation</span>
<span class="sd">is chosen because the fastest-changing dimension as new files are added to the</span>
<span class="sd">collection is usually the cells rather than genes.</span>

<span class="sd">Process:</span>

<span class="sd">1. **Scan the Collection**: Scan the entire collection of files to create</span>
<span class="sd">a unique set of feature ids (e.g. gene symbols). Store this set as the</span>
<span class="sd">`gene_annotation` TileDB file.</span>

<span class="sd">2. **Sample Metadata**: Store sample metadata in `sample_metadata`</span>
<span class="sd">TileDB file. Each file is typically considered a sample, and an automatic</span>
<span class="sd">mapping is created between files and samples.</span>

<span class="sd">3. **Store Cell Metadata**: Store cell metadata in the `cell_metadata`</span>
<span class="sd">TileDB file.</span>

<span class="sd">4. **Remap and Orient Data**: For each dataset in the collection,</span>
<span class="sd">remap and orient the feature dimension using the feature set from Step 1.</span>
<span class="sd">This step ensures consistency in gene measurement and order, even if</span>
<span class="sd">some genes are unmeasured or ordered differently in the original experiments.</span>

<span class="sd">Example:</span>

<span class="sd">    .. code-block:: python</span>

<span class="sd">        import anndata</span>
<span class="sd">        import numpy as np</span>
<span class="sd">        import tempfile</span>
<span class="sd">        from cellarr import build_cellarrdataset, CellArrDataset, MatrixOptions</span>

<span class="sd">        # Create a temporary directory</span>
<span class="sd">        tempdir = tempfile.mkdtemp()</span>

<span class="sd">        # Read AnnData objects</span>
<span class="sd">        adata1 = anndata.read_h5ad(&quot;path/to/object1.h5ad&quot;)</span>
<span class="sd">        # or just provide the path</span>
<span class="sd">        adata2 = &quot;path/to/object2.h5ad&quot;</span>

<span class="sd">        # Build CellArrDataset</span>
<span class="sd">        dataset = build_cellarrdataset(</span>
<span class="sd">            output_path=tempdir,</span>
<span class="sd">            files=[adata1, adata2],</span>
<span class="sd">            matrix_options=MatrixOptions(dtype=np.float32),</span>
<span class="sd">        )</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Union</span>

<span class="kn">import</span> <span class="nn">anndata</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">utils_anndata</span> <span class="k">as</span> <span class="n">uad</span>
<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">buildutils_tiledb_array</span> <span class="k">as</span> <span class="n">uta</span>
<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">buildutils_tiledb_frame</span> <span class="k">as</span> <span class="n">utf</span>
<span class="kn">from</span> <span class="nn">.CellArrDataset</span> <span class="kn">import</span> <span class="n">CellArrDataset</span>
<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">build_options</span> <span class="k">as</span> <span class="n">bopt</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s2">&quot;Jayaram Kancherla&quot;</span>
<span class="n">__copyright__</span> <span class="o">=</span> <span class="s2">&quot;Jayaram Kancherla&quot;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s2">&quot;MIT&quot;</span>


<span class="c1"># TODO: Accept files as a dictionary with names to each dataset.</span>
<div class="viewcode-block" id="build_cellarrdataset">
<a class="viewcode-back" href="../../api/cellarr.html#cellarr.build_cellarrdataset.build_cellarrdataset">[docs]</a>
<span class="k">def</span> <span class="nf">build_cellarrdataset</span><span class="p">(</span>
    <span class="n">files</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">anndata</span><span class="o">.</span><span class="n">AnnData</span><span class="p">]],</span>
    <span class="n">output_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gene_annotation</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="nb">str</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">sample_metadata</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">cell_metadata</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">sample_metadata_options</span><span class="p">:</span> <span class="n">bopt</span><span class="o">.</span><span class="n">SampleMetadataOptions</span> <span class="o">=</span> <span class="n">bopt</span><span class="o">.</span><span class="n">SampleMetadataOptions</span><span class="p">(),</span>
    <span class="n">cell_metadata_options</span><span class="p">:</span> <span class="n">bopt</span><span class="o">.</span><span class="n">CellMetadataOptions</span> <span class="o">=</span> <span class="n">bopt</span><span class="o">.</span><span class="n">CellMetadataOptions</span><span class="p">(),</span>
    <span class="n">gene_annotation_options</span><span class="p">:</span> <span class="n">bopt</span><span class="o">.</span><span class="n">GeneAnnotationOptions</span> <span class="o">=</span> <span class="n">bopt</span><span class="o">.</span><span class="n">GeneAnnotationOptions</span><span class="p">(),</span>
    <span class="n">matrix_options</span><span class="p">:</span> <span class="n">bopt</span><span class="o">.</span><span class="n">MatrixOptions</span> <span class="o">=</span> <span class="n">bopt</span><span class="o">.</span><span class="n">MatrixOptions</span><span class="p">(),</span>
    <span class="n">optimize_tiledb</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="n">num_threads</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate the `CellArrDataset`.</span>

<span class="sd">    All files are expected to be consistent and any modifications</span>
<span class="sd">    to make them consistent is outside the scope of this function</span>
<span class="sd">    and package.</span>

<span class="sd">    There&#39;s a few assumptions this process makes:</span>
<span class="sd">    - If object in ``files`` is an :py:class:`~anndata.AnnData`</span>
<span class="sd">    or H5AD object, these must contain an assay matrix in the</span>
<span class="sd">    layers slot of the object named as ``layer_matrix_name`` parameter.</span>
<span class="sd">    - Feature information must contain a column defined by the parameter</span>
<span class="sd">    ``feature_column`` in the</span>
<span class="sd">    :py:class:`~cellarr.build_options.GeneAnnotationOptions.` that</span>
<span class="sd">    contains feature ids or gene symbols across all files.</span>
<span class="sd">    - If no ``cell_metadata`` is provided, we scan to count the number of cells</span>
<span class="sd">    and create a simple range index.</span>
<span class="sd">    - Each file is considered a sample and a mapping between cells and samples</span>
<span class="sd">    is automatically created. Hence the sample information provided must match</span>
<span class="sd">    the number of input files.</span>

<span class="sd">    Args:</span>
<span class="sd">        files:</span>
<span class="sd">            List of file paths to `H5AD` or ``AnnData`` objects.</span>

<span class="sd">        output_path:</span>
<span class="sd">            Path to where the output tiledb files should be stored.</span>

<span class="sd">        gene_metadata:</span>
<span class="sd">            A :py:class:`~pandas.DataFrame` containing the feature/gene</span>
<span class="sd">            annotations across all objects.</span>

<span class="sd">            Alternatively, may provide a path to the file containing</span>
<span class="sd">            a concatenated gene annotations across all datasets.</span>
<span class="sd">            In this case, the first row is expected to contain the</span>
<span class="sd">            column names and an index column containing the feature ids</span>
<span class="sd">            or gene symbols.</span>

<span class="sd">            Alternatively, a list or a dictionary of gene symbols.</span>

<span class="sd">            Irrespective of the input, the object will be appended</span>
<span class="sd">            with a ``cellarr_gene_index`` column that contains the</span>
<span class="sd">            gene index across all objects.</span>

<span class="sd">            Defaults to None, then a gene set is generated by scanning all</span>
<span class="sd">            objects in ``files``.</span>

<span class="sd">        sample_metadata:</span>
<span class="sd">            A :py:class:`~pandas.DataFrame` containing the sample</span>
<span class="sd">            metadata for each file in ``files``. Hences the number of rows</span>
<span class="sd">            in the dataframe must match the number of ``files``.</span>

<span class="sd">            Alternatively, may provide path to the file containing a</span>
<span class="sd">            concatenated sample metadata across all cells. In this case,</span>
<span class="sd">            the first row is expected to contain the column names.</span>

<span class="sd">            Additionally, the order of rows is expected to be in the same</span>
<span class="sd">            order as the input list of ``files``.</span>

<span class="sd">            Irrespective of the input, this object is appended with a</span>
<span class="sd">            ``cellarr_original_gene_list`` column that contains the original</span>
<span class="sd">            set of feature ids (or gene symbols) from the dataset to</span>
<span class="sd">            differentiate between zero-expressed vs unmeasured genes.</span>

<span class="sd">            Defaults to `None`, in which case, we create a simple sample</span>
<span class="sd">            metadata dataframe containing the list of datasets.</span>
<span class="sd">            Each dataset is named as ``sample_{i}`` where `i` refers to</span>
<span class="sd">            the index position of the object in ``files``.</span>

<span class="sd">        cell_metadata:</span>
<span class="sd">            A :py:class:`~pandas.DataFrame` containing the cell</span>
<span class="sd">            metadata for cells across ``files``. Hences the number of rows</span>
<span class="sd">            in the dataframe must match the number of cells across</span>
<span class="sd">            all files.</span>

<span class="sd">            Alternatively, may provide path to the file containing a</span>
<span class="sd">            concatenated cell metadata across all cells. In this case,</span>
<span class="sd">            the first row is expected to contain the column names.</span>

<span class="sd">            Additionally, the order of cells is expected to be in the same</span>
<span class="sd">            order as the input list of ``files``. If the input is a path,</span>
<span class="sd">            the file is expected to contain mappings between cells and</span>
<span class="sd">            datasets (or samples).</span>

<span class="sd">            Defaults to None, we scan all files to count the number of cells,</span>
<span class="sd">            then create a simple cell metadata DataFrame containing mappings from</span>
<span class="sd">            cells to their associated datasets. Each dataset is named as</span>
<span class="sd">            ``sample_{i}`` where `i` refers to the index position of</span>
<span class="sd">            the object in ``files``.</span>

<span class="sd">        sample_metadata_options:</span>
<span class="sd">            Optional parameters when generating ``sample_metadata`` store.</span>

<span class="sd">        cell_metadata_options:</span>
<span class="sd">            Optional parameters when generating ``cell_metadata`` store.</span>

<span class="sd">        gene_annotation_options:</span>
<span class="sd">            Optional parameters when generating ``gene_annotation`` store.</span>

<span class="sd">        matrix_options:</span>
<span class="sd">            Optional parameters when generating ``matrix`` store.</span>

<span class="sd">        optimize_tiledb:</span>
<span class="sd">            Whether to run TileDb&#39;s vaccum and consolidation (may take long).</span>

<span class="sd">        num_threads:</span>
<span class="sd">            Number of threads.</span>
<span class="sd">            Defaults to 1.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">output_path</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;&#39;output_path&#39; must be a directory.&quot;</span><span class="p">)</span>

    <span class="n">uad</span><span class="o">.</span><span class="n">extract_anndata_info</span><span class="p">(</span>
        <span class="n">files</span><span class="p">,</span>
        <span class="n">var_feature_column</span><span class="o">=</span><span class="n">gene_annotation_options</span><span class="o">.</span><span class="n">feature_column</span><span class="p">,</span>
        <span class="n">num_threads</span><span class="o">=</span><span class="n">num_threads</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1">####</span>
    <span class="c1">## Writing gene annotation file</span>
    <span class="c1">####</span>
    <span class="k">if</span> <span class="n">gene_annotation</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;Scanning all files for feature ids (e.g. gene symbols), this may take long&quot;</span><span class="p">,</span>
            <span class="ne">UserWarning</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">gene_set</span> <span class="o">=</span> <span class="n">uad</span><span class="o">.</span><span class="n">scan_for_features</span><span class="p">()</span>
        <span class="n">gene_set</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gene_set</span><span class="p">)</span>
        <span class="n">gene_annotation</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;cellarr_gene_index&quot;</span><span class="p">:</span> <span class="n">gene_set</span><span class="p">},</span> <span class="n">index</span><span class="o">=</span><span class="n">gene_set</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_annotation</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
        <span class="n">_gene_list</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">gene_annotation</span><span class="p">)))</span>
        <span class="n">gene_annotation</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span><span class="s2">&quot;cellarr_gene_index&quot;</span><span class="p">:</span> <span class="n">_gene_list</span><span class="p">},</span> <span class="n">index</span><span class="o">=</span><span class="n">_gene_list</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_annotation</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">gene_annotation</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gene_annotation</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;Using the index of the DataFrame to collect feature ids or gene symbols...&quot;</span><span class="p">,</span>
            <span class="ne">UserWarning</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">gene_annotation</span><span class="p">[</span><span class="s2">&quot;cellarr_gene_index&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_annotation</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;&#39;gene_annotation&#39; is not an expected type.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">gene_annotation</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;&#39;gene_annotation&#39; must be a pandas dataframe.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">gene_annotation</span><span class="p">[</span><span class="s2">&quot;cellarr_gene_index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span>
        <span class="n">gene_annotation</span><span class="p">[</span><span class="s2">&quot;cellarr_gene_index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;&#39;gene_annotation&#39; must contain unique feature ids or gene symbols.&quot;</span>
        <span class="p">)</span>

    <span class="n">gene_annotation</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Create the gene annotation tiledb</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_annotation_options</span><span class="o">.</span><span class="n">skip</span><span class="p">:</span>
        <span class="n">_col_types</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">gene_annotation</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">_col_types</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;ascii&quot;</span>

        <span class="n">_gene_output_uri</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">gene_annotation_options</span><span class="o">.</span><span class="n">tiledb_store_name</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">generate_metadata_tiledb_frame</span><span class="p">(</span>
            <span class="n">_gene_output_uri</span><span class="p">,</span> <span class="n">gene_annotation</span><span class="p">,</span> <span class="n">column_types</span><span class="o">=</span><span class="n">_col_types</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">optimize_tiledb</span><span class="p">:</span>
            <span class="n">uta</span><span class="o">.</span><span class="n">optimize_tiledb_array</span><span class="p">(</span><span class="n">_gene_output_uri</span><span class="p">)</span>

    <span class="c1">####</span>
    <span class="c1">## Writing the sample metadata file</span>
    <span class="c1">####</span>
    <span class="n">_samples</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">files</span><span class="p">):</span>
        <span class="n">_samples</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;sample_</span><span class="si">{</span><span class="n">idx</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">sample_metadata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;Sample metadata is not provided, each dataset in &#39;files&#39; is considered a sample&quot;</span><span class="p">,</span>
            <span class="ne">UserWarning</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">sample_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;cellarr_sample&quot;</span><span class="p">:</span> <span class="n">_samples</span><span class="p">})</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">sample_metadata</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">sample_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">sample_metadata</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">sample_metadata</span><span class="p">[</span><span class="s2">&quot;cellarr_sample&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_samples</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s2">&quot;&#39;sample_metadata&#39; is not an expected type.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">sample_metadata_options</span><span class="o">.</span><span class="n">skip</span><span class="p">:</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;Scanning all files for feature ids (e.g. gene symbols), this may take long&quot;</span><span class="p">,</span>
            <span class="ne">UserWarning</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">gene_scan_set</span> <span class="o">=</span> <span class="n">uad</span><span class="o">.</span><span class="n">scan_for_features</span><span class="p">(</span><span class="n">unique</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">gene_set_str</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">gene_scan_set</span><span class="p">]</span>
        <span class="n">sample_metadata</span><span class="p">[</span><span class="s2">&quot;cellarr_original_gene_set&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_set_str</span>

        <span class="n">_col_types</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">sample_metadata</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">_col_types</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;ascii&quot;</span>

        <span class="n">_sample_output_uri</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">sample_metadata_options</span><span class="o">.</span><span class="n">tiledb_store_name</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">generate_metadata_tiledb_frame</span><span class="p">(</span>
            <span class="n">_sample_output_uri</span><span class="p">,</span> <span class="n">sample_metadata</span><span class="p">,</span> <span class="n">column_types</span><span class="o">=</span><span class="n">_col_types</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">optimize_tiledb</span><span class="p">:</span>
            <span class="n">uta</span><span class="o">.</span><span class="n">optimize_tiledb_array</span><span class="p">(</span><span class="n">_sample_output_uri</span><span class="p">)</span>

    <span class="c1">####</span>
    <span class="c1">## Writing the cell metadata file</span>
    <span class="c1">####</span>
    <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
        <span class="s2">&quot;Scanning all files to compute cell counts, this may take long&quot;</span><span class="p">,</span>
        <span class="ne">UserWarning</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">cell_counts</span> <span class="o">=</span> <span class="n">uad</span><span class="o">.</span><span class="n">scan_for_cellcounts</span><span class="p">()</span>
    <span class="n">_cellindex_in_dataset</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">_dataset</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">cci</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cell_counts</span><span class="p">):</span>
        <span class="n">_cellindex_in_dataset</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cci</span><span class="p">)])</span>
        <span class="n">_dataset</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;dataset_</span><span class="si">{</span><span class="n">idx</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cci</span><span class="p">)])</span>

    <span class="k">if</span> <span class="n">cell_metadata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cell_metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="p">{</span><span class="s2">&quot;cellarr_cell_counts&quot;</span><span class="p">:</span> <span class="n">_cellindex_in_dataset</span><span class="p">,</span> <span class="s2">&quot;cellarr_sample&quot;</span><span class="p">:</span> <span class="n">_dataset</span><span class="p">}</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cell_metadata</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;Scanning &#39;cell_metadata&#39; csv file to count number of cells, this may take long&quot;</span><span class="p">,</span>
            <span class="ne">UserWarning</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">cell_metadata</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>
            <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">fp</span><span class="p">:</span>
                <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">cell_counts</span><span class="p">)</span> <span class="o">!=</span> <span class="n">count</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Number of rows in the &#39;cell_metadata&#39; csv does not match the number of cells across files.&quot;</span>
            <span class="p">)</span>

        <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span>
            <span class="s2">&quot;&#39;cell_metadata&#39; csv file is expected to contain mapping between cells and samples&quot;</span><span class="p">,</span>
            <span class="ne">UserWarning</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cell_metadata</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">cell_counts</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cell_metadata</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Number of rows in &#39;cell_metadata&#39; does not match the number of cells across files.&quot;</span>
            <span class="p">)</span>

        <span class="n">cell_metadata</span><span class="p">[</span><span class="s2">&quot;cellarr_sample&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_dataset</span>

    <span class="c1"># Create the cell metadata tiledb</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">cell_metadata_options</span><span class="o">.</span><span class="n">skip</span><span class="p">:</span>
        <span class="n">_cell_output_uri</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">/cell_metadata&quot;</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cell_metadata</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">_cell_metaframe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cell_metadata</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">generate_metadata_tiledb_csv</span><span class="p">(</span>
                <span class="n">_cell_output_uri</span><span class="p">,</span> <span class="n">cell_metadata</span><span class="p">,</span> <span class="n">_cell_metaframe</span><span class="o">.</span><span class="n">columns</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cell_metadata</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">):</span>
            <span class="n">_col_types</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">cell_metadata</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">_col_types</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;ascii&quot;</span>

            <span class="n">_to_write</span> <span class="o">=</span> <span class="n">cell_metadata</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

            <span class="n">generate_metadata_tiledb_frame</span><span class="p">(</span>
                <span class="n">_cell_output_uri</span><span class="p">,</span> <span class="n">_to_write</span><span class="p">,</span> <span class="n">column_types</span><span class="o">=</span><span class="n">_col_types</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">optimize_tiledb</span><span class="p">:</span>
            <span class="n">uta</span><span class="o">.</span><span class="n">optimize_tiledb_array</span><span class="p">(</span><span class="n">_cell_output_uri</span><span class="p">)</span>

    <span class="c1">####</span>
    <span class="c1">## Writing the matrix file</span>
    <span class="c1">####</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">matrix_options</span><span class="o">.</span><span class="n">skip</span><span class="p">:</span>
        <span class="n">gene_idx</span> <span class="o">=</span> <span class="n">gene_annotation</span><span class="p">[</span><span class="s2">&quot;cellarr_gene_index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">gene_set</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">gene_idx</span><span class="p">):</span>
            <span class="n">gene_set</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="o">=</span> <span class="n">i</span>

        <span class="n">_counts_uri</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">output_path</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">matrix_options</span><span class="o">.</span><span class="n">tiledb_store_name</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="n">uta</span><span class="o">.</span><span class="n">create_tiledb_array</span><span class="p">(</span>
            <span class="n">_counts_uri</span><span class="p">,</span>
            <span class="n">matrix_attr_name</span><span class="o">=</span><span class="n">matrix_options</span><span class="o">.</span><span class="n">matrix_name</span><span class="p">,</span>
            <span class="n">x_dim_dtype</span><span class="o">=</span><span class="n">cell_metadata_options</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span>
            <span class="n">y_dim_dtype</span><span class="o">=</span><span class="n">gene_annotation_options</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span>
            <span class="n">matrix_dim_dtype</span><span class="o">=</span><span class="n">matrix_options</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">fd</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">mat</span> <span class="o">=</span> <span class="n">uad</span><span class="o">.</span><span class="n">remap_anndata</span><span class="p">(</span>
                <span class="n">fd</span><span class="p">,</span>
                <span class="n">gene_set</span><span class="p">,</span>
                <span class="n">var_feature_column</span><span class="o">=</span><span class="n">gene_annotation_options</span><span class="o">.</span><span class="n">feature_column</span><span class="p">,</span>
                <span class="n">layer_matrix_name</span><span class="o">=</span><span class="n">matrix_options</span><span class="o">.</span><span class="n">matrix_name</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">uta</span><span class="o">.</span><span class="n">write_csr_matrix_to_tiledb</span><span class="p">(</span>
                <span class="n">_counts_uri</span><span class="p">,</span>
                <span class="n">matrix</span><span class="o">=</span><span class="n">mat</span><span class="p">,</span>
                <span class="n">row_offset</span><span class="o">=</span><span class="n">offset</span><span class="p">,</span>
                <span class="n">value_dtype</span><span class="o">=</span><span class="n">matrix_options</span><span class="o">.</span><span class="n">dtype</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">offset</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">mat</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="k">if</span> <span class="n">optimize_tiledb</span><span class="p">:</span>
            <span class="n">uta</span><span class="o">.</span><span class="n">optimize_tiledb_array</span><span class="p">(</span><span class="n">_counts_uri</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">CellArrDataset</span><span class="p">(</span>
        <span class="n">dataset_path</span><span class="o">=</span><span class="n">output_path</span><span class="p">,</span>
        <span class="n">sample_metadata_uri</span><span class="o">=</span><span class="n">sample_metadata_options</span><span class="o">.</span><span class="n">tiledb_store_name</span><span class="p">,</span>
        <span class="n">cell_metadata_uri</span><span class="o">=</span><span class="n">cell_metadata_options</span><span class="o">.</span><span class="n">tiledb_store_name</span><span class="p">,</span>
        <span class="n">gene_annotation_uri</span><span class="o">=</span><span class="n">gene_annotation_options</span><span class="o">.</span><span class="n">tiledb_store_name</span><span class="p">,</span>
        <span class="n">matrix_tdb_uri</span><span class="o">=</span><span class="n">matrix_options</span><span class="o">.</span><span class="n">tiledb_store_name</span><span class="p">,</span>
    <span class="p">)</span></div>



<div class="viewcode-block" id="generate_metadata_tiledb_frame">
<a class="viewcode-back" href="../../api/cellarr.html#cellarr.build_cellarrdataset.generate_metadata_tiledb_frame">[docs]</a>
<span class="k">def</span> <span class="nf">generate_metadata_tiledb_frame</span><span class="p">(</span>
    <span class="n">output_uri</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">input</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">column_types</span><span class="p">:</span> <span class="nb">dict</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate metadata tiledb from a :pu:class:`~pandas.DataFrame`.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_uri:</span>
<span class="sd">            TileDB URI or path to save the file.</span>

<span class="sd">        input:</span>
<span class="sd">            Input dataframe.</span>

<span class="sd">        column_types:</span>
<span class="sd">            You can specify type of each column name to cast into.</span>
<span class="sd">            &quot;ascii&quot; or `str` works best for most scenarios.</span>

<span class="sd">            Defaults to None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_to_write</span> <span class="o">=</span> <span class="nb">input</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">utf</span><span class="o">.</span><span class="n">create_tiledb_frame_from_dataframe</span><span class="p">(</span>
        <span class="n">output_uri</span><span class="p">,</span> <span class="n">_to_write</span><span class="p">,</span> <span class="n">column_types</span><span class="o">=</span><span class="n">column_types</span>
    <span class="p">)</span></div>



<div class="viewcode-block" id="generate_metadata_tiledb_csv">
<a class="viewcode-back" href="../../api/cellarr.html#cellarr.build_cellarrdataset.generate_metadata_tiledb_csv">[docs]</a>
<span class="k">def</span> <span class="nf">generate_metadata_tiledb_csv</span><span class="p">(</span>
    <span class="n">output_uri</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="nb">input</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">column_dtype</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
    <span class="n">chunksize</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate a metadata tiledb from csv.</span>

<span class="sd">    The difference between this and ``generate_metadata_tiledb_frame``</span>
<span class="sd">    is when the csv is super large and it won&#39;t fit into memory.</span>

<span class="sd">    Args:</span>
<span class="sd">        output_uri:</span>
<span class="sd">            TileDB URI or path to save the file.</span>

<span class="sd">        input:</span>
<span class="sd">            Path to the csv file. The first row is expected to</span>
<span class="sd">            contain the column names.</span>

<span class="sd">        column_dtype:</span>
<span class="sd">            Dtype of the columns.</span>
<span class="sd">            Defaults to str.</span>

<span class="sd">        chunksize:</span>
<span class="sd">            Chunk size to read the dataframe.</span>
<span class="sd">            Defaults to 1000.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">chunksize</span> <span class="o">=</span> <span class="mi">1000</span>
    <span class="n">initfile</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">offset</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">chunk</span> <span class="ow">in</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="n">chunksize</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">initfile</span><span class="p">:</span>
            <span class="n">utf</span><span class="o">.</span><span class="n">create_tiledb_frame_from_column_names</span><span class="p">(</span>
                <span class="n">output_uri</span><span class="p">,</span> <span class="n">chunk</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">column_dtype</span>
            <span class="p">)</span>
            <span class="n">initfile</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="n">_to_write</span> <span class="o">=</span> <span class="n">chunk</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">utf</span><span class="o">.</span><span class="n">append_to_tiledb_frame</span><span class="p">(</span><span class="n">output_uri</span><span class="p">,</span> <span class="n">_to_write</span><span class="p">,</span> <span class="n">offset</span><span class="p">)</span>
        <span class="n">offset</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">chunk</span><span class="p">)</span></div>

</pre></div>
        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          
          
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Jayaram Kancherla
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="../../_static/documentation_options.js?v=47de8214"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/furo.js?v=4e2eecee"></script>
    </body>
</html>